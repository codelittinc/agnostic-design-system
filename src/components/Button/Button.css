@import "../../styles/theme/index.css";

.button {
  background-color: darkcyan;
  border: none;
  color: white;
  padding: 0 var(--button-primary-large-side-padding);
}
