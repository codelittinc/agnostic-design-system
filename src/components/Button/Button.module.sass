.button
  background: linear-gradient(90deg, #FF465D 0%, #BC46BA 100%)
  box-shadow: 0 0.125rem 0.125rem rgba(0, 0, 0, 0.2)
  border-radius: 0.25rem
  border: none
  color: #fff
  padding: 0.75rem 1.5rem
  outline: none
  cursor: pointer
  box-sizing: border-box
  font-size: 1.125rem
  font-weight: 400
  transition: box-shadow 100ms linear
  display: flex
  align-items: center
  justify-content: center
  min-height: 3rem

  &:hover:not([disabled])
    box-shadow: 0 0.5rem 1rem rgba(0, 0, 0, 0.2)

  &:active:not([disabled])
    box-shadow: 0 0.125rem 0.125rem rgba(0, 0, 0, 0.2), inset 0 0 2rem rgba(0, 0, 0, 0.3)

  &[disabled]
    box-shadow: none
    background: #E9ECEC
    color: #C8CED0

  &.big
    padding: 1.25rem 1.5rem
    min-height: 4rem

  &.full-width
    width: 100%

  &.icon-left
    flex-direction: row
    > span
      margin-left: 1rem

  &.icon-right
    flex-direction: row-reverse
    > span
      margin-right: 1rem

  &.icon-full-left
    flex-direction: row
    padding-left: 1rem
    > span
      margin-left: 1rem
      flex: 1

  &.icon-full-right
    flex-direction: row-reverse
    padding-right: 1rem
    > span
      margin-right: 1rem
      flex: 1
