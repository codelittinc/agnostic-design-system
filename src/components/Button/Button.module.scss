@import '~@codelitt/ay-css-library/src/components/brand/fonts';
@import '~@codelitt/ay-css-library/src/components/brand/colors';

.button {
  border-radius: 0.125rem;
  border: 0;
  font-family: $font-stack;
  font-size: 1.25rem;
  height: 3.5rem;
  width: 16rem;
}

.main {
  background-color: $ay-london-blue-color;
  &:disabled {
    background-color: $ay-grey-8-color;
    color: $ay-grey-24-color;
  }

  &:not(:disabled):focus {
    background-color: $ay-navy-blue-color;
    color: $ay-white-color;
  }

  &:not(:disabled):hover {
    background-color: $ay-main-blue-color;
    color: $ay-white-color;
  }
}

