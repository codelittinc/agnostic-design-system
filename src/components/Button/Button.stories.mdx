# Button Component

## Button Props

|        Prop        |      Type       | Default | Optional |
| :----------------: | :-------------: | :-----: | :------: |
|     appendIcon     | React.ReactNode |    -    |   true   |
|      content       | React.ReactNode |    -    |   true   |
|      disabled      |     boolean     |  false  |   true   |
|       label        |     string      |    -    |   true   |
|      category      |    Category     | neutral |   true   |
|       fixed        |     boolean     |  false  |   true   |
|      onClick       |   () => void    |    -    |  false   |
|        size        |      Size       |  large  |   true   |
| variablesClassName |     string      |    -    |   true   |
|   withAppendIcon   |     boolean     |  false  |   true   |

---

|     Props      | Description                                                                                                                                                                                                                                                                                                                                                                                                                                                                    |
| :------------: | :----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| **appendIcon** | By default it receives an [Arrow Icon](https://design-system.codelitt.dev/?path=/story/components-button--with-append-icon), but it is possible to send your own [Icon](https://design-system.codelitt.dev/?path=/story/components-button--with-custom-append-icon). The Icon can also be the [only element inside the button](https://design-system.codelitt.dev/?path=/story/components-button--without-label-with-append-icon), if there isn't a value passed to the label. |

**Categories availables:**

1.  [neutral](https://design-system.codelitt.dev/?path=/story/components-button--neutral)
2.  [primary](https://design-system.codelitt.dev/?path=/story/components-button--primary)
3.  [secondary](https://design-system.codelitt.dev/?path=/story/components-button--secondary)
4.  [positive](https://design-system.codelitt.dev/?path=/story/components-button--positive)
5.  [negative](https://design-system.codelitt.dev/?path=/story/components-button--negative)
6.  [ghost](https://design-system.codelitt.dev/?path=/story/components-button--ghost)

**Sizes**

1. large
2. medium

## Getting Started

```javascript
import React, { Component } from 'react';
import { Button } from '@codelittinc/agnostic-design-system';

class SimpleButton extends Component {
  const handleClick = () => {
    console.log('Button clicked!');
  }

  render() {
    return (
      <Button onClick={handleClick} />
    );
  }
}

export default SimpleButton;

```

# Customazing styles

For the Button component we use the following naming convention to configure the CSS variables:

**--button-[property]?-[attribute]-[state]?**

- Property - append-icon (optional)
- Attribute - background-color, border, font, etc (required)
- State - hover, focus, active, disabled (optional)

1. Create the css class using our variables

```css
.my-button-css {
  /* Component + Attribute */
  --button-background-color: red; /* Simplest variable */

  /* Component + Attribute + State */
  --button-background-color-hover: blue; /* Using state */

  /* Component + Property + Attribute */
  --button-append-icon-color: green;

  /* Component + Property + Attribute + State */
  --button-append-icon-color-active: red;
}
```

2. Pass the css class through its `variablesClassName` property

`<Button variablesClassName={styles['my-button-css']} />`

## Variables available

|  Property   |    Attribute     |             State              |
| :---------: | :--------------: | :----------------------------: |
|             |   align-items    |                                |
|             | background-color | hover, focus, active, disabled |
|             |      border      | hover, focus, active, disabled |
|             |  border-radius   |                                |
|             |    box-shadow    | hover, focus, active, disabled |
|             |      color       | hover, focus, active, disabled |
|             |      cursor      | hover, focus, active, disabled |
|             |     display      |                                |
|             |   font-family    |                                |
|             |    font-size     |                                |
|             |   font-weight    |                                |
|             |      height      |                                |
|             |   line-height    |                                |
|             |      margin      |                                |
|             |     padding      |                                |
| append-icon |      color       | hover, focus, active, disabled |
| append-icon |      margin      |                                |
