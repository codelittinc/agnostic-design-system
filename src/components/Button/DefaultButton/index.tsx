import React from 'react';
import classNames from 'classnames';
import styles from '@/components/Button/DefaultButton/DefaultButton.css';
import DefaultAppendIcon from '@/assets/images/icons/web/append-button-icon.svg';

interface Props {
  appendIcon?: React.ReactNode;
  className: string;
  content?: React.ReactNode;
  disabled?: boolean;
  fixed?: boolean;
  onClick?: () => void;
  size?: string;
  text?: string;
  type?: 'button' | 'submit' | 'reset' | undefined;
  variablesClassName?: string;
  withAppendIcon?: boolean;
}

const getAppendIcon = (appendIcon, size) => {
  const largeIcon = size === 'large';
  const sizeWidthIcon = largeIcon ? 8 : 6;
  const sizeHeightIcon = largeIcon ? 14 : 10;

  return appendIcon ? (
    <>{appendIcon}</>
  ) : (
    <DefaultAppendIcon
      className={classNames(styles['button--append-icon'])}
      height={sizeHeightIcon}
      width={sizeWidthIcon}
    />
  );
};

const getButtonContent = (text, withAppendIcon, appendIcon, size) => {
  return (
    <>
      {text ? <span className={classNames(styles['button--text'])}>{text}</span> : ''}

      {withAppendIcon && (
        <div className={classNames(styles['button--append-icon-container'])}>
          {getAppendIcon(appendIcon, size)}
        </div>
      )}
    </>
  );
};

const DefaultButton: React.FC<Props> = props => {
  const {
    appendIcon,
    className,
    content,
    fixed,
    onClick,
    size,
    text,
    type,
    variablesClassName,
    withAppendIcon,
    ...buttonProps
  } = props;

  const sizeClass = `button--${size}`;

  const buttonContent = content || getButtonContent(text, withAppendIcon, appendIcon, size);
  return (
    <button
      {...buttonProps}
      onClick={onClick}
      type={type || 'button'}
      className={classNames(
        className,
        styles.button,
        styles[sizeClass],
        {
          [styles['button--fixed']]: fixed
        },
        variablesClassName
      )}
    >
      {buttonContent}
    </button>
  );
};

DefaultButton.defaultProps = {
  disabled: false,
  fixed: false,
  size: 'large'
};

export default DefaultButton;
