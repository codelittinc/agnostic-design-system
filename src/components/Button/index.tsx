import React from 'react';

import PrimaryButton from '@/components/Button/PrimaryButton';
import SecondaryButton from '@/components/Button/SecondaryButton';

type Category = 'primary' | 'secondary';

interface Props {
  disabled?: boolean;
  label?: string;
  category?: Category,
  onClick: () => void;
}

const buttons = {
  'primary': PrimaryButton,
  'secondary': SecondaryButton,
};

const Button: React.FC<Props> = props => {
  const ButtonType = buttons[props.category || 'primary'];

  return (
    <ButtonType {...props} />
  );
};

export default Button;
