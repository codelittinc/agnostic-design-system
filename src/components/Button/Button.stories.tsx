import React from 'react';
import { storiesOf } from '@storybook/react';
import '../../global.css';
import { text, boolean, withKnobs, select } from '@storybook/addon-knobs';
import { action } from '@storybook/addon-actions';
import Button from './index';
import exampleIcon from '../../assets/images/icons/icon.svg';

const stories = storiesOf('Components/Button', module);

stories.addDecorator(withKnobs);

stories.add('Text', () => {
  const buttonText = text('text', 'Click me');
  const wrapperClassName = text('wrapperClassName', '');
  const disabled = boolean('disabled', false);
  const big = boolean('big', false);
  const fullWidth = boolean('fullWidth', false);

  return (
    <Button
      onClick={action('button-click')}
      text={buttonText}
      wrapperClassName={wrapperClassName}
      disabled={disabled}
      big={big}
      fullWidth={fullWidth}
    />
  );
});

stories.add('Text and icon', () => {
  const buttonText = text('text', 'Click me');
  const wrapperClassName = text('wrapperClassName', '');
  const disabled = boolean('disabled', false);
  const big = boolean('big', false);
  const fullWidth = boolean('fullWidth', false);

  const iconPositionOptions: any = ['left', 'right', 'full-left', 'full-right'];
  const iconPosition = select('iconPosition', iconPositionOptions, 'left');

  return (
    <Button
      icon={exampleIcon}
      iconPosition={iconPosition}
      onClick={action('button-click')}
      text={buttonText}
      wrapperClassName={wrapperClassName}
      disabled={disabled}
      big={big}
      fullWidth={fullWidth}
    />
  );
});

stories.add('Only icon', () => {
  const wrapperClassName = text('wrapperClassName', '');
  const disabled = boolean('disabled', false);
  const big = boolean('big', false);
  const fullWidth = boolean('fullWidth', false);

  return (
    <Button
      icon={exampleIcon}
      onClick={action('button-click')}
      wrapperClassName={wrapperClassName}
      disabled={disabled}
      big={big}
      fullWidth={fullWidth}
    />
  );
});
