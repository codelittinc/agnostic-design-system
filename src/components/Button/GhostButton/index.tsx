import React from 'react';
import DefaultButton from '@/components/Button/DefaultButton';
import classNames from 'classnames';

import styles from '@/components/Button/GhostButton/GhostButton.css';

interface Props {
  disabled?: boolean;
  fixed?: boolean,
  label?: string;
  onClick: () => void;
}

const GhostButton: React.FC<Props> = props => {
  return (
    <DefaultButton
      {...props}
      className={classNames(styles.button, styles['button--ghost'], styles['button--large'])}
    />
  );
};

export default GhostButton;
