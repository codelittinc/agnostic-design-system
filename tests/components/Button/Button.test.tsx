import React from 'react';
import { cleanup, fireEvent, render } from '@testing-library/react';
import Button, {
  Type,
  IconPosition,
  Size,
} from '../../../src/components/Button';

afterEach(cleanup);

describe('Button', () => {
  const getComponent = (
    type: Type = 'main',
    onClick: () => void,
    iconPosition?: IconPosition,
    size?: Size,
  ) =>
    render(
      <Button
        id="1"
        name="button1"
        label="Button Label"
        onClick={onClick}
        type={type}
        disabled={false}
        iconPosition={iconPosition}
        size={size}
      />,
    );

  it('renders without crashing', () => {
    expect(getComponent('main', () => {})).toMatchSnapshot();
  });

  it('calls onClick prop on button click', () => {
    let clickCount = 0;

    const addClick = () => {
      clickCount += 1;
    };

    const { getByText } = getComponent('main', addClick);
    const node = getByText('Button Label');

    fireEvent.click(node);
    expect(clickCount).toBe(1);
  });

  it('renders a button with a left icon', () => {
    const { getByText } = getComponent('main', () => {}, 'left');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button main left-icon regular-size',
    );
  });

  it('renders a button with a right icon', () => {
    const { getByText } = getComponent('main', () => {}, 'right');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button main right-icon regular-size',
    );
  });

  it('renders a button with small size', () => {
    const { getByText } = getComponent('main', () => {}, undefined, 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty('className', 'button main  small-size');
  });

  it('renders a button with small size and a left icon', () => {
    const { getByText } = getComponent('main', () => {}, 'left', 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button main left-icon small-size',
    );
  });

  it('renders a button with small size and a right icon', () => {
    const { getByText } = getComponent('main', () => {}, 'right', 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button main right-icon small-size',
    );
  });

  it('renders a secondary button', () => {
    const { getByText } = getComponent('secondary', () => {});
    const node = getByText('Button Label');
    expect(node).toHaveProperty('className', 'button secondary  regular-size');
  });

  it('renders a secondary button with a left icon', () => {
    const { getByText } = getComponent('secondary', () => {}, 'left');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button secondary left-icon regular-size',
    );
  });

  it('renders a secondary button with a right icon', () => {
    const { getByText } = getComponent('secondary', () => {}, 'right');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button secondary right-icon regular-size',
    );
  });

  it('renders a secondary button with small size', () => {
    const { getByText } = getComponent(
      'secondary',
      () => {},
      undefined,
      'small',
    );
    const node = getByText('Button Label');
    expect(node).toHaveProperty('className', 'button secondary  small-size');
  });

  it('renders a secondary button with small size and a left icon', () => {
    const { getByText } = getComponent('secondary', () => {}, 'left', 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button secondary left-icon small-size',
    );
  });

  it('renders a secondary button with small size and a right icon', () => {
    const { getByText } = getComponent('secondary', () => {}, 'right', 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button secondary right-icon small-size',
    );
  });

  it('renders a ghost button', () => {
    const { getByText } = getComponent('ghost', () => {});
    const node = getByText('Button Label');
    expect(node).toHaveProperty('className', 'button ghost  regular-size');
  });

  it('renders a ghost button with a left icon', () => {
    const { getByText } = getComponent('ghost', () => {}, 'left');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button ghost left-icon regular-size',
    );
  });

  it('renders a ghost button with a right icon', () => {
    const { getByText } = getComponent('ghost', () => {}, 'right');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button ghost right-icon regular-size',
    );
  });

  it('renders a ghost button with small size', () => {
    const { getByText } = getComponent('ghost', () => {}, undefined, 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty('className', 'button ghost  small-size');
  });

  it('renders a ghost button with small size and a left icon', () => {
    const { getByText } = getComponent('ghost', () => {}, 'left', 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button ghost left-icon small-size',
    );
  });

  it('renders a ghost button with small size and a right icon', () => {
    const { getByText } = getComponent('ghost', () => {}, 'right', 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button ghost right-icon small-size',
    );
  });

  it('renders a contextual button', () => {
    const { getByText } = getComponent('contextual', () => {});
    const node = getByText('Button Label');
    expect(node).toHaveProperty('className', 'button contextual  regular-size');
  });

  it('renders a contextual button with a left icon', () => {
    const { getByText } = getComponent('contextual', () => {}, 'left');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button contextual left-icon regular-size',
    );
  });

  it('renders a contextual button with a right icon', () => {
    const { getByText } = getComponent('contextual', () => {}, 'right');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button contextual right-icon regular-size',
    );
  });

  it('renders a contextual button with small size', () => {
    const { getByText } = getComponent(
      'contextual',
      () => {},
      undefined,
      'small',
    );
    const node = getByText('Button Label');
    expect(node).toHaveProperty('className', 'button contextual  small-size');
  });

  it('renders a contextual button with small size and a left icon', () => {
    const { getByText } = getComponent('contextual', () => {}, 'left', 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button contextual left-icon small-size',
    );
  });

  it('renders a contextual button with small size and a right icon', () => {
    const { getByText } = getComponent(
      'contextual',
      () => {},
      'right',
      'small',
    );
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button contextual right-icon small-size',
    );
  });

  it('renders a delete button', () => {
    const { getByText } = getComponent('delete', () => {});
    const node = getByText('Button Label');
    expect(node).toHaveProperty('className', 'button delete  regular-size');
  });

  it('renders a delete button with a left icon', () => {
    const { getByText } = getComponent('delete', () => {}, 'left');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button delete left-icon regular-size',
    );
  });

  it('renders a delete button with a right icon', () => {
    const { getByText } = getComponent('delete', () => {}, 'right');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button delete right-icon regular-size',
    );
  });

  it('renders a delete button with small size', () => {
    const { getByText } = getComponent('delete', () => {}, undefined, 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty('className', 'button delete  small-size');
  });

  it('renders a delete button with small size and a left icon', () => {
    const { getByText } = getComponent('delete', () => {}, 'left', 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button delete left-icon small-size',
    );
  });

  it('renders a delete button with small size and a right icon', () => {
    const { getByText } = getComponent('delete', () => {}, 'right', 'small');
    const node = getByText('Button Label');
    expect(node).toHaveProperty(
      'className',
      'button delete right-icon small-size',
    );
  });
});
