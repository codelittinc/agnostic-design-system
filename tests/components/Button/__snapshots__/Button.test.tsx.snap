// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExampleComponent is renders correctly 1`] = `
<button
  className="button"
  onClick={[Function]}
  title="this is a test"
  type="button"
>
  <span>
    this is a test
  </span>
</button>
`;
